expression/builtin_time.go

```diff
@@ -975,13 +975,13 @@ type builtinDayOfMonthSig struct {
 // evalInt evals a builtinDayOfMonthSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_dayofmonth
 func (b *builtinDayOfMonthSig) evalInt(row types.Row) (int64, bool, error) {
+	oldMode := b.ctx.GetSessionVars().SQLMode
+	b.ctx.GetSessionVars().SQLMode = oldMode & (^mysql.ModeNoZeroInDate)
```

Comment:
According to https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_dayofmonth,  date value like `2008-00-00` is valid for builtin function `dayofmonth`, so sql mode `NO_ZERO_IN_DATE` has to be canceld when calculating. It's somewhat ugly way here, I think the thorough solution is to refine the time check & sql mode handling system, it's something big. @zz-jason

Reply:
in mysql:
```sql
MySQL > select dayOfMonth('2017-00-00');
+--------------------------+
| dayOfMonth('2017-00-00') |
+--------------------------+
|                        0 |
+--------------------------+
1 row in set (0.01 sec)

MySQL > select dayOfMonth('0000-00-00');
+--------------------------+
| dayOfMonth('0000-00-00') |
+--------------------------+
|                     NULL |
+--------------------------+
1 row in set, 1 warning (0.01 sec)

MySQL > show warnings;
+---------+------+----------------------------------------+
| Level   | Code | Message                                |
+---------+------+----------------------------------------+
| Warning | 1292 | Incorrect datetime value: '0000-00-00' |
+---------+------+----------------------------------------+
1 row in set (0.01 sec)
```

Reply:
@zz-jason  yes. That's what I do: cancelling `NO_ZERO_IN_DATE`, reserve `NO_ZERO_DATE`.  Your examples acts the same with my change. PTAL